urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/IOHelper.java

```diff
@@ -0,0 +1,215 @@
+		v |= ((((long) buf[offset + 7]) & 0xFF) << 0);
+		return v;
+	}
+
+	public static int intIdFromString(final String in) {
+		try {
+			final MessageDigest md = MessageDigest.getInstance(MD_ALG);
+			final byte[] b = md.digest(in.getBytes(UTF8));
+			return (IOHelper.intFromByteArray(b, 0) ^ IOHelper.intFromByteArray(b, 4)
+					^ IOHelper.intFromByteArray(b, 8) ^ IOHelper.intFromByteArray(b, 12))
+					& Integer.MAX_VALUE;
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static long longIdFromString(final String in) {
+		try {
+			final MessageDigest md = MessageDigest.getInstance(MD_ALG);
+			final byte[] b = md.digest(in.getBytes(UTF8));
+			return (IOHelper.longFromByteArray(b, 0) ^ IOHelper.longFromByteArray(b, 8)) & Long.MAX_VALUE;
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static final void toWireWithHeader(final OutputStream os, final byte[] buf, final int len)
+			throws IOException {
+		final byte[] header = new byte[4]; // Integer
+		if (len > 0xFFFF) { // Limit to 64KB
+			throw new IOException("Packet length overflow (" + len + ")");
+		}
+		intToByteArray((len & 0xFFFF) | LENGTH_MAGIC, header, 0);
+		os.write(header, 0, header.length);
+		os.write(buf, 0, len);
+		os.flush();
+	}
+
+	public static final byte[] fromWireWithHeader(final InputStream is) throws IOException {
+		final byte[] header = new byte[4]; // Integer
+		int readed = -1;
+		readed = fullRead(is, header, header.length);
+		if (readed <= 0) {
+			throw new EOFException("EOF");
+		}
+		if (readed != header.length) {
+			throw new IOException("Invalid HEADER");
+		}
+		int len = intFromByteArray(header, 0);
+		if ((len & 0xFFFF0000) != LENGTH_MAGIC) {
+			throw new IOException("Invalid MAGIC");
+		}
+		len &= 0xFFFF; // Limit to 64KB
+		if (len > (Constants.BUFFER_LEN << 1)) {
+			throw new IOException("Packet length overflow (" + len + ")");
+		}
+		final byte[] buf = new byte[len];
+		readed = fullRead(is, buf, buf.length);
+		if (readed != buf.length) {
+			throw new IOException("Invalid BODY");
+		}
+		return buf;
+	}
+
+	public static void setupSocket(final ServerSocket sock) throws SocketException {
+		sock.setReuseAddress(true);
+		sock.setSoTimeout(Constants.ACCEPT_TIMEOUT);
+		sock.setReceiveBufferSize(Math.max(sock.getReceiveBufferSize(), Constants.BUFFER_LEN
+				* Constants.IO_BUFFERS));
+	}
+
+	public static void setupSocket(final Socket sock) throws SocketException {
+		sock.setKeepAlive(true);
+		sock.setReuseAddress(true);
+		sock.setSoTimeout(Constants.READ_TIMEOUT);
+		sock.setSendBufferSize(Math.max(sock.getSendBufferSize(), Constants.BUFFER_LEN * Constants.IO_BUFFERS));
+		sock.setReceiveBufferSize(Math.max(sock.getReceiveBufferSize(), Constants.BUFFER_LEN
+				* Constants.IO_BUFFERS));
+	}
+
+	public static void closeSilent(final Reader ir) {
+		if (ir == null)
+			return;
+		try {
+			ir.close();
+		} catch (Exception ign) {
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 154:159, 163:168, 204:209.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.